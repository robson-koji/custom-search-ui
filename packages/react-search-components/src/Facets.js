import PropTypes from "prop-types";
import React from "react";

function Facets({ children }) {
  return <div className="facets">{children}</div>;
}

Facets.propTypes = {
  children: PropTypes.node.isRequired
};

export default Facets;
