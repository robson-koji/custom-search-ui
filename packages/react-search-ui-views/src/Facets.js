import PropTypes from "prop-types";
import React from "react";

import { appendClassName } from "./view-helpers";

function Facets({ children, className }) {
  return (
    <div className={appendClassName("sui-facet-container", className)}>
      {children}
    </div>
  );
}

Facets.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string
};

export default Facets;
