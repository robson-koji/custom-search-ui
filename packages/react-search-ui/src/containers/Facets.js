import PropTypes from "prop-types";
import React, { Component } from "react";
import { withSearch } from "..";
import { Facet, Facets } from "@elastic/react-search-components";

import { FacetDetail, Filter } from "../types";

function capitalizeFirstLetter(string) {
  return string.charAt(0).toUpperCase() + string.slice(1);
}

function findFacetValueInFilters(name, filters) {
  const filter = filters.find(f => Object.keys(f)[0] === name);
  if (!filter) return;
  return Object.values(filter)[0][0];
}

export class FacetsContainer extends Component {
  static propTypes = {
    addFilter: PropTypes.func.isRequired,
    filters: PropTypes.arrayOf(Filter).isRequired,
    facets: PropTypes.objectOf(FacetDetail).isRequired,
    removeFilter: PropTypes.func.isRequired
  };

  render() {
    const { addFilter, filters, facets, removeFilter } = this.props;
    const facetNames = Object.keys(facets);
    if (!facetNames.length) return null;

    return (
      <Facets>
        {facetNames.map(name => {
          const options = facets[name][0].data;
          const selectedValue = findFacetValueInFilters(name, filters);
          if (!options.length && !selectedValue) return null;

          return (
            <Facet
              key={name}
              name={capitalizeFirstLetter(name)}
              onRemove={({ clickEvent, value }) => {
                clickEvent.preventDefault();
                removeFilter(name, value);
              }}
              onSelect={({ clickEvent, value }) => {
                clickEvent.preventDefault();
                addFilter(name, value);
              }}
              options={options}
              value={selectedValue}
            />
          );
        })}
      </Facets>
    );
  }
}

export default withSearch(FacetsContainer);
