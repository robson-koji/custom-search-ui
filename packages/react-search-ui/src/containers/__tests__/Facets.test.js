import React from "react";
import { shallow } from "enzyme";
import { FacetsContainer } from "../Facets";

const params = {
  addFilter: jest.fn(),
  filters: [{ field: ["value"] }],
  facets: {
    field: [
      {
        data: [],
        type: "value"
      }
    ],
    another: [
      {
        data: [{ count: 10, value: "anotherValue" }],
        type: "value"
      }
    ],
    empty: [
      {
        data: [],
        type: "value"
      }
    ]
  },
  removeFilter: jest.fn()
};

beforeEach(() => {
  params.addFilter = jest.fn();
  params.removeFilter = jest.fn();
});

it("renders correctly", () => {
  const wrapper = shallow(<FacetsContainer {...params} />);
  expect(wrapper).toMatchSnapshot();
});

it("should not render a Facet if there are no options available", () => {
  const wrapper = shallow(<FacetsContainer {...params} />);
  // There are three facets
  expect(Object.keys(params.facets).length).toEqual(3);
  // But only 2 are rendered
  expect(wrapper.find("Facet").length).toEqual(2);
});

it("will call back when a facet value is selected", () => {
  const wrapper = shallow(<FacetsContainer {...params} />);

  wrapper
    .find("Facet")
    .at(1)
    .prop("onSelect")({
    clickEvent: {
      preventDefault: () => {}
    },
    value: "anotherValue"
  });

  const [fieldName, fieldValue] = params.addFilter.mock.calls[0];
  expect(fieldName).toEqual("another");
  expect(fieldValue).toEqual("anotherValue");
});

it("will call back when a facet value is removed", () => {
  const wrapper = shallow(<FacetsContainer {...params} />);

  wrapper
    .find("Facet")
    .at(0)
    .prop("onRemove")({
    clickEvent: {
      preventDefault: () => {}
    },
    value: "value"
  });

  const [fieldName, fieldValue] = params.removeFilter.mock.calls[0];
  expect(fieldName).toEqual("field");
  expect(fieldValue).toEqual("value");
});
