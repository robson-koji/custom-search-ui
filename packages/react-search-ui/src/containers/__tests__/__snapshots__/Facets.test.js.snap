// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FacetsContainer
    addFilter={[MockFunction]}
    facets={
      Object {
        "another": Array [
          Object {
            "data": Array [
              Object {
                "count": 10,
                "value": "anotherValue",
              },
            ],
            "type": "value",
          },
        ],
        "empty": Array [
          Object {
            "data": Array [],
            "type": "value",
          },
        ],
        "field": Array [
          Object {
            "data": Array [],
            "type": "value",
          },
        ],
      }
    }
    filters={
      Array [
        Object {
          "field": Array [
            "value",
          ],
        },
      ]
    }
    removeFilter={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Facet
          name="Field"
          onRemove={[Function]}
          onSelect={[Function]}
          options={Array []}
          value="value"
        />,
        <Facet
          name="Another"
          onRemove={[Function]}
          onSelect={[Function]}
          options={
            Array [
              Object {
                "count": 10,
                "value": "anotherValue",
              },
            ]
          }
        />,
        null,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "field",
        "nodeType": "function",
        "props": Object {
          "name": "Field",
          "onRemove": [Function],
          "onSelect": [Function],
          "options": Array [],
          "value": "value",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "another",
        "nodeType": "function",
        "props": Object {
          "name": "Another",
          "onRemove": [Function],
          "onSelect": [Function],
          "options": Array [
            Object {
              "count": 10,
              "value": "anotherValue",
            },
          ],
          "value": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      null,
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Facet
            name="Field"
            onRemove={[Function]}
            onSelect={[Function]}
            options={Array []}
            value="value"
          />,
          <Facet
            name="Another"
            onRemove={[Function]}
            onSelect={[Function]}
            options={
              Array [
                Object {
                  "count": 10,
                  "value": "anotherValue",
                },
              ]
            }
          />,
          null,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "field",
          "nodeType": "function",
          "props": Object {
            "name": "Field",
            "onRemove": [Function],
            "onSelect": [Function],
            "options": Array [],
            "value": "value",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "another",
          "nodeType": "function",
          "props": Object {
            "name": "Another",
            "onRemove": [Function],
            "onSelect": [Function],
            "options": Array [
              Object {
                "count": 10,
                "value": "anotherValue",
              },
            ],
            "value": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        null,
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
